File: src/game/systems/setcurrentanimation.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Animation = require 'game.components.animation'
local Movement = require 'game.components.movement'

local aspect = Aspect.new({Animation, Movement})
local SetCurrentAnimation = System:new('setcurrentanimation', aspect)

function SetCurrentAnimation:update(dt)
  for _, entity in pairs(self.entities) do
    local animation = entity:as(Animation)
    local movement = entity:as(Movement)

    local action = 'stand'

    if movement.right or movement.left then
      action = 'walk'
    end

    local newAnimation = action .. '_' .. movement.direction

    if newAnimation ~= animation.currentAnimation then
      animation.currentAnimation = newAnimation
      animation.currentFrame = 1
    end
  end
end

return SetCurrentAnimation

File: packages/client/src/game/systems/SetCurrentAnimation.ts
import { System } from '../../ecs/System'
import { Aspect } from '../../ecs/Aspect'
import { SystemFlag } from '../flags'
import { Animation } from '../components/Animation'
import { Movement } from '../components/Movement'

export class SetCurrentAnimationSystem extends System {
  static _id = 'SetCurrentAnimation'
  _id = SetCurrentAnimationSystem._id

  static _flag = SystemFlag.SetCurrentAnimation
  _flag = SetCurrentAnimationSystem._flag

  static _aspect = new Aspect([Animation, Movement])
  _aspect = SetCurrentAnimationSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const animation = entity.as(Animation)
      const movement = entity.as(Movement)

      if (!animation || !movement) {
        return
      }

      let action = 'stand'

      if (movement.right || movement.left) {
        action = 'walk'
      }

      const newAnimation = `${action}_${movement.direction}`

      // if (newAnimation !== animation.currentSequence) {
      // animation.currentAnimation = newAnimation
      //   animation.currentFrame = 1
      // }
    })
  }
}
